### PR TITLE
Possible fix for #3886

### PR DESCRIPTION
**Pull request by adamdickmeiss**
pendingEvents quene introduced.

Test case missing for this PR.

### CODE REVIEW

**Comment by vietj** [comment]
we need a test for this

**Comment by adamdickmeiss** [comment]
The test case reveals a more fundamental problem with the stream
that feeds JsonParser in #3886. It's as it does not respect
pause and will continue to feed. So it seems this is not
a solution but a work-around may use a lot of memory.

**Comment by vietj** [comment]
can you provide a reproducer for the non respected pause ?

**Comment by adamdickmeiss** [comment]
> can you provide a reproducer for the non respected pause ?

I'll try.. I added a flag in JsonParserImpl to see if my assertion was correct. It seems not.. Ie it does seem to be able to occur without that claim. I guess that means the problem might still be in JsonParserImpl and not in the stream being used.

**Comment by adamdickmeiss** [comment]
Got it now.. stream==null in original case and that causes trouble.

File: src/main/java/io/vertx/core/parsetools/impl/JsonParserImpl.java [hunk 1]
```diff
@@ -221,6 +232,16 @@ public void end() {
...
+      while (demand > 0L && !pendingEvents.isEmpty()) {
+        if (demand != Long.MAX_VALUE) {
+          demand--;
+        }
+        JsonEvent jsonEvent = pendingEvents.removeFirst();
+        Handler<JsonEvent> handler = this.eventHandler;
+        if (handler != null) {
+          handler.handle(jsonEvent);
+        }
+      }
```
Comment (julianladisch) [inline src/main/java/io/vertx/core/parsetools/impl/JsonParserImpl.java:244]: The while loop should be inside the try clause. If the eventHandler throws an exception we want `finally` `emitting = false;`

Comment (adamdickmeiss) [reply src/main/java/io/vertx/core/parsetools/impl/JsonParserImpl.java]: Right

**Comment by vietj** [comment]
were do we stand now @adamdickmeiss ?

File: src/main/java/io/vertx/core/parsetools/impl/JsonParserImpl.java [hunk 2]
```diff
@@ -72,15 +75,18 @@ public JsonParser pause() {
 
   @Override
   public JsonParser resume() {
-    return fetch(Long.MAX_VALUE);
+    demand = Long.MAX_VALUE;
+    checkPending();
```
Comment (vietj) [inline src/main/java/io/vertx/core/parsetools/impl/JsonParserImpl.java:79]: why do we need this change ?

File: src/main/java/io/vertx/core/parsetools/impl/JsonParserImpl.java [hunk 3]
```diff
@@ -114,6 +120,10 @@ public JsonParser handler(Handler<JsonEvent> handler) {
   }
 
   private void handleEvent(JsonEvent event) {
+    if (demand == 0L) {
+      pendingEvents.addLast(event);
+      return;
```
Comment (vietj) [inline src/main/java/io/vertx/core/parsetools/impl/JsonParserImpl.java:125]: can you avoid return and instead use else-if ?

**Comment by vietj** [comment]
@adamdickmeiss I believe the internal of the current `JsonParser` to rely on the internal parser to handle unconsumed tokens is flawed. I refactored the `JsonParser` to consume all tokens and buffer events instead (like your PR do) and went for more drastic changes in https://github.com/eclipse-vertx/vert.x/pull/3912

**Comment by adamdickmeiss** [comment]
If PR #3912 solves the problem of #3886 then we're good and this can be closed (duplicate work)

**Comment by vietj** [comment]
can you review code in the PR ?

**Comment by adamdickmeiss** [comment]
> can you review code in the PR ?
> […](#)

OK

**Comment by vietj** [comment]
thanks for the reproducer and analysis @adamdickmeiss it definitely helped a lot

**Pull request closed without merging**